providers/siw/siw.c

```diff
@@ -0,0 +1,941 @@
+	if (!mr)
+		return NULL;
+
+	rv = ibv_cmd_reg_mr(pd, addr, len, (uintptr_t)addr, access,
+			    &mr->base_mr, &cmd.base, sizeof(cmd), &resp.base,
+			    sizeof(resp));
+	if (rv) {
+		free(mr);
+		return NULL;
+	}
+	return &mr->base_mr.ibv_mr;
+}
+
+static int siw_dereg_mr(struct verbs_mr *base_mr)
+{
+	struct siw_mr *mr = mr_base2siw(base_mr);
+	int rv;
+
+	rv = ibv_cmd_dereg_mr(base_mr);
+	if (rv)
+		return rv;
+
+	free(mr);
+	return 0;
+}
+
+static struct ibv_cq *siw_create_cq(struct ibv_context *ctx, int num_cqe,
+				    struct ibv_comp_channel *channel,
+				    int comp_vector)
+{
+	struct siw_cmd_create_cq cmd;
+	struct siw_cmd_create_cq_resp resp;
+	struct siw_cq *cq;
+	int rv;
+
+	memset(&cmd, 0, sizeof(cmd));
+	memset(&resp, 0, sizeof(resp));
+
+	cq = calloc(1, sizeof(*cq));
+	if (!cq)
+		return NULL;
+
+	rv = ibv_cmd_create_cq(ctx, num_cqe, channel, comp_vector, &cq->base_cq,
+			       &cmd.base, sizeof(cmd), &resp.base,
+			       sizeof(resp));
+	if (rv) {
+		if (siw_debug)
+			printf("libsiw: CQ creation failed: %d\n", rv);
+		free(cq);
+		return NULL;
+	}
+	pthread_spin_init(&cq->lock, PTHREAD_PROCESS_PRIVATE);
+
+	cq->id = resp.siw.cq_id;
+	cq->num_cqe = resp.siw.num_cqe;
+
+	if (resp.siw.cq_key != SIW_INVAL_UOBJ_KEY) {
+		int cq_size = resp.siw.num_cqe * sizeof(struct siw_cqe) +
+			      sizeof(struct siw_cq_ctrl);
+
+		cq->queue = mmap(NULL, cq_size, PROT_READ | PROT_WRITE,
+				 MAP_SHARED, ctx->cmd_fd, resp.siw.cq_key);
+
+		if (cq->queue == MAP_FAILED) {
+			if (siw_debug)
+				printf("libsiw: CQ mapping failed: %d", errno);
+			goto fail;
+		}
+		cq->ctrl = (struct siw_cq_ctrl *)&cq->queue[cq->num_cqe];
+		cq->ctrl->notify = SIW_NOTIFY_NOT;
+
+		return &cq->base_cq;
+	}
+	if (siw_debug)
+		printf("libsiw: prepare CQ mapping failed\n");
+fail:
+	ibv_cmd_destroy_cq(&cq->base_cq);
+	free(cq);
+
+	return NULL;
+}
+
+static int siw_resize_cq(struct ibv_cq *base_cq, int num_cqe)
+{
+	return -EOPNOTSUPP;
+}
+
+static int siw_destroy_cq(struct ibv_cq *base_cq)
+{
+	struct siw_cq *cq = cq_base2siw(base_cq);
+	int rv;
+
+	pthread_spin_lock(&cq->lock);
+
+	if (cq->queue)
+		munmap(cq->queue, cq->num_cqe * sizeof(struct siw_cqe) +
+					  sizeof(struct siw_cq_ctrl));
+
+	rv = ibv_cmd_destroy_cq(base_cq);
+	if (rv) {
+		pthread_spin_unlock(&cq->lock);
+		return rv;
+	}
+	pthread_spin_unlock(&cq->lock);
+
+	free(cq);
+
+	return 0;
+}
+
+static struct ibv_srq *siw_create_srq(struct ibv_pd *pd,
+				      struct ibv_srq_init_attr *attr)
+{
+	struct siw_cmd_create_srq cmd;
+	struct siw_cmd_create_srq_resp resp;
+	struct siw_srq *srq;
+
+	memset(&cmd, 0, sizeof(cmd));
+	memset(&resp, 0, sizeof(resp));
+
+	srq = calloc(1, sizeof(*srq));
+	if (!srq)
+		return NULL;
+
+	if (ibv_cmd_create_srq(pd, &srq->base_srq, attr, &cmd.base, sizeof(cmd),
+			       &resp.base, sizeof(resp))) {
+		free(srq);
+		return NULL;
+	}
+	pthread_spin_init(&srq->lock, PTHREAD_PROCESS_PRIVATE);
+
+	if (resp.siw.srq_key != SIW_INVAL_UOBJ_KEY) {
```

Comment:
success oriented flow, here and in other places like this

Reply:
OK will do that for all similar places as well...